### ISSUE TITLE
Keep track of Access and Refresh Tokens from OIDC flows

### ISSUE THREAD

**Issue by aeneasr** [ory/kratos#397]
**Is your feature request related to a problem? Please describe.**

We should store access/refresh tokens retrieved from oidc flows. Potentially in the credentials struct?

This also requires us to define a way to retrieve it and/or use Kratos as a proxy when calling e.g. GitHub APIs.

**Comment by sash2222** [comment]
Hello @aeneasr ! This is exactly what we need, we want to use this token further in our service, so as not to force the user to log in again further inside the service. Right now, I can't allocate a GO programmer for this task, but more than likely we can connect to its implementation along the way.
As for the implementation, the token may well be harnessed in the user data structure, knowing the UUID, we can pull it out of the API. In our case, it will be used by a backend and it might be better to store it, for example, in the Hashicorp Vault, this is almost an industry standard.

**Comment by aeneasr** [comment]
Yeah that makes sense - integration with Vault. I think we can have multiple strategies - the first one being to store it (encrypted) in the database. Then using the identity ID we can look it up at the admin API. Another strategy would be vault :)

**Comment by sash2222** [comment]
Damn the right words! I am in favor of both options, in the general mass of course the API is a more important approach and possibly simpler at this stage. But in the strategic Vault, it's great, it's cloudnative :).

**Comment by aeneasr** [comment]
Plus it doesn't add external dependency requirements ;)

**Comment by johanforssell** [comment]
I'm trying to pick out the AD groups from a Microsoft OIDC login, as traits. As soon as there are more than 5 groups associated with a user you need to fetch them yourself from the Microsoft Graph API.

Of course - connecting to the Graph API you also need the tokens - getting them from the API would certainly make things easier for me.

**Comment by aeneasr** [comment]
We‘d definitely welcome contributions in that area!

**Comment by david972** [comment]
Hello there, I have started to implement vault feature before seeing this issue in order to store data like access_token.
But in my implementation I did not plan to create a other strategy

I planned to integrate the field with the type, the secret path into identity

I'm really interested by what you are thinking about this implementation

**Comment by aeneasr** [comment]
In the easiest scenario we just store the initial access and refresh token somewhere (e.g. the credentials), in a more complex one we always return a valid access token (by refreshing them internally)

**Comment by aeneasr** [comment]
You mean Hashicorp Vault?

**Comment by david972** [comment]
yes

**Comment by aeneasr** [comment]
We currently don't want to add more external dependencies to get Ory Kratos running. The vault solution only really makes sense when Vault is in your stack already. Adopting Vault because you want access tokens from Ory Kratos would be cumbersome for most ops :/

**Comment by david972** [comment]
🤔 ok I understand

**Comment by david972** [comment]
What is your advise / idea ? I can change my way

**Comment by aeneasr** [comment]
https://github.com/ory/kratos/issues/397#issuecomment-856525353 <- what I said here

It would probably make sense to encrypt the tokens as well at rest

**Comment by david972** [comment]
Actually, my first thought was to do it like that, so
Ok i'll do it like that

**Comment by david972** [comment]
hello
Should I create a new handler or update the credential struct to add the access_token and refresh token ?
If the new handler is preferable I suggest that it check the access_token validity and refresh it if it’s not and render the result What do you think ?

**Comment by aeneasr** [comment]
Looks like you [missed my reply on slack](https://ory-community.slack.com/archives/C012USDT5QQ/p1624876017023800) :)

> I think it should be a separate handler. Maybe you can draft a proposal and post it in the issue so that we can discuss it there?

**Comment by sash2222** [comment]
Friends, it's great that the work is going on. I have a couple of questions:
1. Do I understand correctly that this token can be used from an external system by requesting Kratos in the API and receiving it from the user profile in clear text, or will it be necessary to decrypt it on the side of the microservice?
2. Is it possible to provide for the ability to save other tokens in the Kratos user profile through the API? A scenario when the backend somehow receives this token and wants to keep it as close to the user as possible.

**Comment by david972** [comment]
1. Yes the goal is to retrieve the token in clear the token will be encrypted before recording into database
2. It isn't plan to do it for now for what I know.

**Comment by aeneasr** [comment]
Yeah option 2 really isn’t in scope of Ory Kratos - something like hashicorp vault might be better suited.

@david972 lmk when good for 👀:)

**Comment by david972** [comment]
actually I waiting for your second review to add the test

**Comment by aeneasr** [comment]
Ah ok, sorry! Tomorrow I have no backlog and will review :)

**Comment by david972** [comment]
do not apologise, there is no problem
